app/connectors/CircuitBreakers.scala

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package connectors
+
+import akka.pattern.CircuitBreaker
+import akka.stream.Materializer
+import config.CircuitBreakerConfig
+import play.api.Logging
+
+import scala.concurrent.duration.Duration
+
+trait CircuitBreakers { self: Logging =>
+  def materializer: Materializer
+  def gbCircuitBreakerConfig: CircuitBreakerConfig
+  def niCircuitBreakerConfig: CircuitBreakerConfig
+
+  private val clazz = getClass.getSimpleName
+
+  lazy val gbCircuitBreaker = new CircuitBreaker(
+    scheduler = materializer.system.scheduler,
+    maxFailures = gbCircuitBreakerConfig.maxFailures,
+    callTimeout = gbCircuitBreakerConfig.callTimeout,
+    resetTimeout = gbCircuitBreakerConfig.resetTimeout,
+    maxResetTimeout = gbCircuitBreakerConfig.maxResetTimeout,
+    exponentialBackoffFactor = gbCircuitBreakerConfig.exponentialBackoffFactor,
+    randomFactor = gbCircuitBreakerConfig.randomFactor
+  )(materializer.executionContext)
+    .onOpen(logger.error(s"GB Circuit breaker for ${clazz} opening due to failures"))
+    .onHalfOpen(logger.warn(s"GB Circuit breaker for ${clazz} resetting after failures"))
```

Comment:
The warnings are for state changes - `onHalfOpen` (CB is letting one call through to see if it succeeds) and `onClose` (said call succeeded so it'll let all calls through again). 

Any missed/failed messages will log as an error.

Reply:
@RobertLaverick "half-open" means it is making a trial connection to see whether the connectivity has been restored rather than failing fast when any requests are made